### PR TITLE
Fixes #18804 - replace inline helps with label helps

### PR DESCRIPTION
**Pull request by ares**
Depends on changes from #4354 
Also for the reviewer, I think we should give enough time to plugin devs to align their code bases. While I can work on some, it's out of my capacity to fix everything. Releasing this with 1.15 introduces a risk that we'd have mixed icons in single form (e.g. rex adding flag to interface, see example below). Therefore I'd suggest to keep this for 1.16

nics

### CODE REVIEW

File: app/helpers/form_helper.rb [hunk 1]
```diff
@@ -304,16 +304,24 @@ def help_inline(inline, error)
   def add_label options, f, attr
     label_size = options.delete(:label_size) || "col-md-2"
     required_mark = check_required(options, f, attr)
-    label = options[:label] == :none ? '' : options.delete(:label)
-    label ||= ((clazz = f.object.class).respond_to?(:gettext_translation_for_attribute_name) &&
-        s_(clazz.gettext_translation_for_attribute_name attr)) if f
-    label = label.present? ? label_tag(attr, "#{label}#{required_mark}", :class => label_size + " control-label") : ''
+    if options[:label] == :none
```
Comment (tbrisker) [inline app/helpers/form_helper.rb]: I wonder, why do we even go into this function if `options[:label] == :none`? all we get is an empty string in that case, and in fact the field helper has another option, `option[:no_label]` that does exactly that.

File: app/helpers/form_helper.rb [hunk 2]
```diff
@@ -304,16 +304,24 @@ def help_inline(inline, error)
...
+      label = ''.html_safe
+    else
+      label = ''.html_safe + options.delete(:label)
+      label += ((clazz = f.object.class).respond_to?(:gettext_translation_for_attribute_name) &&
+        s_(clazz.gettext_translation_for_attribute_name attr).html_safe) if f && label.empty?
+    end
+
+    if options[:label_help].present?
+      label += ' '.html_safe + popover("", options[:label_help], options[:label_help_options] || {})
```
Comment (tbrisker) [inline app/helpers/form_helper.rb]: `html_safe` isn't needed here on the empty space.

File: app/helpers/form_helper.rb [hunk 3]
```diff
@@ -304,16 +304,24 @@ def help_inline(inline, error)
...
+    end
+    label = label.present? ? label_tag(attr, label.to_s + required_mark.to_s, :class => label_size + " control-label") : ''
     label
   end
 
   def check_required options, f, attr
     required = options.delete(:required) # we don't want to use html5 required attr so we delete the option
-    return ' *' if required.nil? ? is_required?(f, attr) : required
+    return ' *'.html_safe if required.nil? ? is_required?(f, attr) : required
```
Comment (tbrisker) [inline app/helpers/form_helper.rb:322]: this doesn't need `html_safe`

File: app/views/lookup_keys/_fields.html.erb [hunk 4]
```diff
@@ -14,36 +14,37 @@
...
+                     :label_help => _("Hide all values for this parameter."),
                      :size => "col-md-1", :label_size => "col-md-2", :table_field => true,
                      :disabled => (is_param && !f.object.override)) %>
     </div>
   </fieldset>
   <fieldset>
     <%= collapsing_header _("Optional input validator"), "#optional_input_validators_#{f.object.id}", "collapsed" %>
     <div id="optional_input_validators_<%= f.object.id %>" class="collapse out">
-      <%= alert :class => 'alert-info', :header => _('Note'), :text => _('If ERB is used in a parameter value, the validation of the value will happen during the ENC request. If the value is invalid, the ENC request will fail.') %>
+      <h6><%= _('If ERB is used in a parameter value, the validation of the value will happen during the ENC request. If the value is invalid, the ENC request will fail.') %></h6>
```
Comment (tbrisker) [inline app/views/lookup_keys/_fields.html.erb:44]: Is this change related?

File: app/views/lookup_keys/_fields.html.erb [hunk 5]
```diff
@@ -53,17 +54,18 @@
...
-                      :title => _("The order in which values are resolved.")).html_safe %> <%= _("Set the order in which values are resolved.") %></h6>
-      <%= textarea_f f, :path, :rows => :auto, :label => _("Order"), :id => 'order', :size => "col-md-8", :onchange => 'fill_in_matchers()', :value => f.object.path, :class => "no-stretch" %>
+      <h6><%= _("Set the order in which values are resolved.") %></h6>
+      <%= textarea_f f, :path, :rows => :auto, :label => _("Order"), :id => 'order', :size => "col-md-8", :onchange => 'fill_in_matchers()', :value => f.object.path, :class => "no-stretch",
+                        :label_help => _("The order in which matchers keys are processed, first match wins.<br> You may use multiple attributes as a matcher key, for example, an order of <code>host group, environment</code> would expect a matcher such as <code>hostgroup = \"web servers\", environment = production</code>").html_safe,
+                        :label_help_options => { :title => _("The order in which values are resolved.") } %>
```
Comment (tbrisker) [inline app/views/lookup_keys/_fields.html.erb]: Does it make sense to have a title for the popover that is almost identical to the header for the section?

File: app/views/nic/_base_form.html.erb [hunk 6]
```diff
@@ -4,24 +4,21 @@
...
+              :label_help => _("Device identifier for this interface. This may be different on various platforms and environments, here are some common examples.<br/><ul>" +
                                 "<li>Use the basic name for physical interface identifiers, e.g. <strong>eth0</strong> or <strong>em0</strong> with biosdevname.</li>" +
                                 "<li>For virtual interfaces, use either alias notation (<strong>eth0:1</strong>, name:index) or VLAN notation (<strong>eth0.15</strong>, name.tag).</li>" +
-                                "<li>For bonds it's common to use <strong>bond0</strong> on Linux, <strong>lagg0</strong> on FreeBSD systems.</li></ul>"),
-                      :title => _("Device identifier"),
-                      :rel => 'popover-modal',
-                      :'data-placement' => "top"
-              ),
+                                "<li>For bonds it's common to use <strong>bond0</strong> on Linux, <strong>lagg0</strong> on FreeBSD systems.</li></ul>").html_safe,
+              :label_help_options => { :title => _("Device identifier"), :rel => 'popover-modal', :'data-placement' => "top" },
```
Comment (tbrisker) [inline app/views/nic/_base_form.html.erb]: title here also seems redundent

File: app/views/nic/bmcs/_bmc.html.erb [hunk 7]
```diff
@@ -3,15 +3,15 @@
 <%= content_tag :span, :id => 'bmc_fields' do %>
   <%= text_f f, :username, :size => "col-md-8", :label_size => "col-md-3" %>
   <%= password_f f, :password, :unset => action_name == "edit", :size => "col-md-8", :label_size => "col-md-3",
-                    :help_inline => popover('',
-                        _("The BMC password will be used by Foreman to access the host's BMC controller via a BMC-enabled smart proxy, if available.") + ' ' +
+                    :label_help => _("The BMC password will be used by Foreman to access the host's BMC controller via a BMC-enabled smart proxy, if available.<br>").html_safe + ' ' +
```
Comment (tbrisker) [inline app/views/nic/bmcs/_bmc.html.erb]: nitpick: should be `<br/>`, and no need for extra space when breaking line.

**Review by tbrisker — CHANGES_REQUESTED** [review]
Looks good in general, see inline comments.

File: app/helpers/form_helper.rb (hunk 1 above)
Comment (ares) [reply app/helpers/form_helper.rb]: maybe to get a required mark? anyway I'd like to avoid bigger helpers refactoring in this PR :-)

Comment (tbrisker) [reply app/helpers/form_helper.rb]: nope, follow the existing code - if it is none then this function just returns an empty string after checking a couple of options.

Comment (tbrisker) [reply app/helpers/form_helper.rb]: If you want to not touch other functions you can short-circuit at the top and `return '' if options[:label] == :none` as the first line so nothing changes otherwise

File: app/helpers/form_helper.rb (hunk 3 above)
Comment (ares) [reply app/helpers/form_helper.rb:322]: That's true only because this helper is called from `add_label` where we append to safe buffer, I think it's a good practice that helper returns safe strings (if they are safe), at least internal rails helpers do. While we don't do it anywhere in core code, I think this is useful.

File: app/views/lookup_keys/_fields.html.erb (hunk 4 above)
Comment (ares) [reply app/views/lookup_keys/_fields.html.erb:44]: No, but it brings consistency with the role form too, it's the same thing that @rohoover suggested there.

File: app/views/lookup_keys/_fields.html.erb (hunk 5 above)
Comment (ares) [reply app/views/lookup_keys/_fields.html.erb]: I'm happy to remove (even though I just kept how it was before) as long as you won't made me to extract `<h6>` change into separate PR :-)

File: app/helpers/form_helper.rb (hunk 2 above)
Comment (ares) [reply app/helpers/form_helper.rb]: it is, or I need to write it as `label = label + ' '.html_safe + popover("", options[:label_help], options[:label_help_options] || {})`

File: app/helpers/form_helper.rb (hunk 1 above)
Comment (ares) [reply app/helpers/form_helper.rb]: updated

**Comment by ares** [comment]
Thanks @tbrisker, once you confirm, I'll squash two commits and update the message so it closes both issues.

**Review by tbrisker — APPROVED** [review]

**Comment by tbrisker** [comment]
LGTM, squash away and I will merge once tests are green.

**Comment by ares** [comment]
Squashed, I closed the original PR.

**Comment by tbrisker** [comment]
Thanks @ares ! Please update the foreman-dev list about this so plugin authors can take advantage of it.

**Comment by dLobatog** [comment]
Awesome work here!

**Comment by ares** [comment]
Thanks @tbrisker, sent with additional PRs to most used plugins.
